### DISCUSSION TITLE
Implementing custom model with object-detection example

### DISCUSSION THREAD

**Discussion by semroai** [AxisCommunications/acap-native-sdk-examples#159] [Q&A] [answered]
Hi,

I'm trying to implement my own object detection model to my Axis P1375 camera. I've been using the object-detection example as a framework (MobileNetV2 w/ COCO), but I can't get it to work. My model has an input of 320x320 (EfficientDet1) which is reflected on the manifest cpu file. I've also changed the tensor sizes accordingly. However, when I try to run the app I get a few errors in the logs: "setupLarod: Unable to load model: Could not load model: Could not verify and build model from buffer."

Using journalctl -u larod via SSH I get an error saying: "The model is not a valid Flatbuffer buffer. Session 9: Could not verify and build model from buffer.

I'm a bit at a loss as to what else I have to modify to make my model work. Any help would be greatly appreciated!

**Comment by Corallo** [comment]
Hi Semorai 
Could you try to run on your device
`larod-client -g /path/to/model -c tflite-cpu`
This way, we can make sure that the issue is with the model and not with the code to begin with.

If it fails, it means your model is for some reason non-compatible. 
It is hard for us to say what is wrong with the model without looking at it.
You could try to load it on netron.app and compare it with the SSD mobilenet that we have in our example and see if you notice something odd. (Loops, quantize and de-quantize operations, custom operations etc).

**Reply by semroai** [reply]
Hi, thanks a lot of the response. This is the error that I get.

![Screenshot 2023-09-20 160956](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/132890288/2147084e-e786-4638-a23d-0d1183ac0a3f)

I'm trying to load this model in particular: https://github.com/semroai/bds/blob/main/bird320x320.tflite

**Reply by Corallo** [reply]
My bad, it was `cpu-tflite`, please try again

**Reply by semroai** [reply]
![image](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/132890288/6db932c5-ad21-41ce-9fc9-a18abf2bbe4e)

No problem, here it is

**Reply by Corallo** [reply]
The model needs to be on the camera, and you have to replace /path/to/model with its location.
You can use `scp` to upload the model on the camera.
Example:
On your computer:
`scp ./<your_model_name>.tflite root@<ip_camera>:/tmp`
On the camera
`larod-client -g /tmp/<your_model_name>.tfilte -c cpu-tflite`

**Reply by semroai** [reply]
Oh sorry my bad, I did not include the path to my model. i just copied your command. 
![image](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/132890288/bbb1d96e-e4c6-4a6b-9839-25493d648f72)

Edit: I've loaded the model up again, and the error is gone. It says model loaded.

**Reply by semroai** [reply]
Hi @Corallo,

Sorry about before. I figured out why my model was not loading right. Thank you very much for the help. I've now been able to load the model into an app and run in on the camera, however, now I've encountered other errors. The log states "Empty JPEG image (DNL not supported)" among other things, so I'm assuming the app is not capturing the images properly. Could it be related to resolution configuration? I'm attaching a text file with the logs. I'd really appreciate your help.
[object-detction-log.txt](https://github.com/AxisCommunications/acap-native-sdk-examples/files/12683767/object-detction-log.txt)

**Reply by Corallo** [reply]
The output of your model must have different format.
From your log I see it is returning: [0.703125,0.148438,0.101562,0.097656]
Our example expects the model output as [top, left bottom, right]
In your case, it doesn't seem the case, as top > bottom.

You'll have to check the output format of your model and adapt the post-processing of the bounding box accordingly 
https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/object-detection/app/object_detection.c#L872

**Reply by semroai** [reply]
Hi @Corallo, thanks a lot for the help. I managed to get the detection working. I'm now trying to interface my object detection app with the axis video motion detection. I would like to trigger my object detection app only when motion is detected in a certain area, and then shut it off. Do you have any recommendations into how to implement this? 

Edit: Ideally I would like to use the events system to have as a first condition the video motion detection and as a second condition a positive output on the object detection. However, this doesn't look possible on the menu.

**Reply by Corallo** [reply]
Hi @semroai 
You can add something like the [axis_event example](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/axevent/subscribe_to_event) to your ACAP to receive the event generated by Axis Video Motion detection.
In short, you can use VMD to select the area you are interested in, then create an event that is triggered by VMD, and your ACAP can subscribe to it as shown in the example. 
Another solution could be to implement the Motion detection in the ACAP directly, we have an [example](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/using-opencv) for that too, but it would require more tweaking

**Reply by semroai** [reply]
Thanks a lot for the suggestions @Corallo. Is there any way to modify my application to send event (as in the send_event.c example (register to the event sysem), and make it show up as a possible condition in the device events rules menu?

**Reply by Corallo** [reply]
Yes, that should be possible, after you create your event like shown in the sent_event.c it will appear in the menu after you start the ACAP.

**Reply by semroai** [reply]
@Corallo, great. Is there something in particular that needs to be in the code for it to show up on as a possible condition? I've quickly loaded up the send_event application, but it doesn't show up as a possible condition after running it. Perhaps it needs something else to be initialized or triggered.

**Reply by semroai** [reply]
Hi @Corallo, I'm using this example to set up and declare a stateful event to the camera: https://axiscommunications.github.io/acap-documentation/docs/api/src/api/axevent/html/ax_event_property_state_declaration_example_8c-example.html

However, I can't seem to figure out how to toggle the state of the key-value set "enabled". The idea would be to toggle that state to true when an object has been detected and to false when nothing is being detected. I'd really appreciate it if you could share how to achieve this in the example. I was unable to get it to work using the documentation.

**Reply by Corallo** [reply, marked as answer]
Hello @semroai 

We are currently working on the send_event example to make show how the event can be crated and be visible from the event system in th GUI.

We might take a bit longer to publish a fix, in the meantime you can try this:

Instead of setting up the event like currently showned in the example [L129-146](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/axevent/send_event/app/send_event.c#L129)
Try the following:
```
    ax_event_key_value_set_add_key_values(
        key_value_set, &error, "topic0", "tnsaxis", "CameraApplicationPlatform",
        AX_VALUE_TYPE_STRING, "topic1", "tnsaxis", "Event_example", AX_VALUE_TYPE_STRING, NULL);

    ax_event_key_value_set_add_key_value(key_value_set, "topic2", "tnsaxis", "event",
                                         AX_VALUE_TYPE_STRING, NULL);
    ax_event_key_value_set_add_nice_names(key_value_set, "topic2", "tnsaxis", "event",
                                          "Event_example:Event", NULL);
    ax_event_key_value_set_add_key_value(key_value_set, "value", NULL, &start_value,
                                         AX_VALUE_TYPE_DOUBLE, NULL);
    ax_event_key_value_set_mark_as_data(key_value_set, "value", NULL, NULL);
```

then, to trigger the event:

[L56](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/axevent/send_event/app/send_event.c#L56)
change the way it is done in the example with:
```
g_autoptr(GError) err = NULL;
   ax_event_key_value_set_add_key_values(key_value_set, &err, "value", NULL, &app_data->value,
                                          AX_VALUE_TYPE_DOUBLE, NULL);
```

Try first building the send_event application, with these changes.
You should see that after starting the applicaiton, a new condition will appear in the GUI:
![image](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/28478841/2424acf2-ba3b-4298-b601-ab8b686dc1e6)

You can for instance try to make the camera LED blink when this event is triggered.
Note: This event carries a double value, you can still use it just as a trigger in the GUI regardless of the value that it carries.

If this works, you can then try to port the send_event to your object detection example.

**Reply by semroai** [reply]
Hi @Corallo, thanks a lot for the information. I've tried to implement your changes, but the application won't start. It immediately stops running. However, I've managed to get the condition to appear using the example I mentioned in my previous comment. I just can't figure out how to toggle the "enabled" condition (key-value set) the right way. What is the correct procedure for changing a value in a set? 

[Example](https://axiscommunications.github.io/acap-documentation/docs/api/src/api/axevent/html/ax_event_property_state_declaration_example_8c-example.html#:~:text=%22enabled%22%2C%20%22tnsaxis%22%2C%20%26enabled%2C%20AX_VALUE_TYPE_BOOL%2C%20NULL)

My process would be something like this:

Set up and declare event with default value FALSE.

- Object is detected:
      - Change event value to TRUE
- Object is not detected:
      - Change event value to FALSE

Edit: I got the modifications to work on the example. Sorry about that.

**Reply by Corallo** [reply]
Hi @semroai 
Also in the case of that other example, it should be enough to use

```
gboolean state = TRUE;
ax_event_key_value_set_add_key_value(set, "enabled", NULL, &state,
                                         AX_VALUE_TYPE_BOOL, NULL);
```
To send an event

**Reply by semroai** [reply]
Thanks a lot @Corallo, I got it to work.
